### PR TITLE
feature/UAR-1136 Get all Corporate Trustees for an OE from Oracle query API

### PR DESCRIPTION
**Pull request by RobertBromfield**
### Description 

Created the endpoint `overseas-entity/trusts/{trustId}/corporate-trustees` to retrieve the trustee data when given the ID of a particular Trust

### Ticket Number 

[UAR-1136](https://companieshouse.atlassian.net/browse/UAR-1136)

### CODE REVIEW

File: src/main/java/uk/gov/ch/model/update/CorporateTrusteeData.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+    @Column(name = "LEGAL_FORM")
+    @JsonProperty("legalForm")
+    private String legalForm;
+
+    @Column(name = "COUNTRY")
+    @JsonProperty("country")
+    private String country;
+
+    @Column(name = "MODIFICATION_ID")
```
Comment (bbell-companieshouse) [inline src/main/java/uk/gov/ch/model/update/CorporateTrusteeData.java:40]: Theres a few fields here not needed, looking at the query there should be 11 fields returned but in this file there are 14

File: src/main/java/uk/gov/ch/service/update/impl/CorporateTrusteesServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @Autowired
+    private CorporateTrusteesRepository corporateTrusteesRepository;
+
+    @Override
+    public List<CorporateTrusteeData> getCorporateTrusteeData(String trustId)
+            throws TrustDataCountNotFoundException {
+        List<CorporateTrusteeData> corporateTrusteeDataList = corporateTrusteesRepository.getCorporateTrustees(
+                trustId);
+
+        DataMap dataMap = new DataMap.Builder().build();
```
Comment (bbell-companieshouse) [inline src/main/java/uk/gov/ch/service/update/impl/CorporateTrusteesServiceImpl.java:31]: I would maybe move this logging into the else block below, just in case corporateTrusteeDataList is null when you call corporateTrusteeDataList.size()

File: src/test/java/uk/gov/ch/service/update/impl/CorporateTrusteesServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+        when(corporateTrusteesRepository.getCorporateTrustees(TRUST_ID)).thenReturn(trustees);
+        List<CorporateTrusteeData> result = corporateTrusteesService.getCorporateTrusteeData(TRUST_ID);
+
+        assertEquals(TRUSTEE_1, result.get(0));
+        assertEquals(TRUSTEE_2, result.get(1));
+    }
+
+    @Test
+    @DisplayName("Get corporate trustees - no trustees returned")
```
Comment (bbell-companieshouse) [inline src/test/java/uk/gov/ch/service/update/impl/CorporateTrusteesServiceImplTest.java:81]: Add an additional test for null data perhaps (apologies i know it wasnt originally in the spike PR)

File: src/main/java/uk/gov/ch/service/update/impl/CorporateTrusteesServiceImpl.java (hunk 2 above)
Comment (RobertBromfield) [reply src/main/java/uk/gov/ch/service/update/impl/CorporateTrusteesServiceImpl.java:31]: If you believe that's possible then a `NullPointerException` will be raised on the if statement instead 

Either way, it will have the same result.

It may be best to just put a null check above if you think a null is likely

File: README.md [hunk 4]
```diff
@@ -14,27 +14,28 @@ In order to run the API locally you'll need the following installed on your mach
...
-**GET**   | `/company/{companyNumber}/corporate-body-appointments/persons-of-significant-control`| Calls service to retrieve the psc data for the company
-**GET**   | `/company/{companyNumber}/confirmation-statement/paid`                               | Calls service to verify confirmation statement payment has been made for company on due date
-**GET**   | `/company/{companyNumber}/register/location`                                         | Calls service to retrieve the registers locations for a company
-**GET**   | `/company/{companyNumber}/registered-email-address`                                  | Calls service to retrieve the registered email address for a company
-**GET**   | `/overseas-entity/{companyNumber}/managing-officers`                                 | Calls service to retrieve the managing officers for an overseas entity
-**GET**   | `/overseas-entity/{companyNumber}/beneficial-owners`                                 | Calls service to retrieve the beneficial owner data for an overseas entity
+ Method   | Path                                                         | Description
```
Comment (AndreKramer7654) [inline README.md:17]: please revert this re-format

File: README.md [hunk 5]
```diff
@@ -14,27 +14,28 @@ In order to run the API locally you'll need the following installed on your mach
...
+ **GET**  | `/company/{companyNumber}/corporate-body-appointments/persons-of-significant-control` | Calls service to retrieve the psc data for the company       
+ **GET**  | `/company/{companyNumber}/confirmation-statement/paid`       | Calls service to verify confirmation statement payment has been made for company on due date 
+ **GET**  | `/company/{companyNumber}/register/location`                 | Calls service to retrieve the registers locations for a company 
+ **GET**  | `/company/{companyNumber}/registered-email-address`          | Calls service to retrieve the registered email address for a company 
+ **GET**  | `/overseas-entity/{companyNumber}/managing-officers`         | Calls service to retrieve the managing officers for an overseas entity 
+ **GET**  | `/overseas-entity/{companyNumber}/beneficial-owners`         | Calls service to retrieve the beneficial owner data for an overseas entity 
+ **GET**  | `overseas-entity/trusts/{trustId}/corporate-trustees`        | Calls service to retrieve the Trustee Data when given the ID of a Trust
```
Comment (AndreKramer7654) [inline README.md:38]: just line 38 to add

File: src/main/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesController.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            OracleQueryApplication.APPLICATION_NAME_SPACE);
+    @Autowired
+    private CorporateTrusteesService corporateTrusteesService;
+
+    @GetMapping("overseas-entity/trusts/{trustId}/corporate-trustees")
+    public ResponseEntity<List<CorporateTrusteeData>> getCorporateTrusteeData(
+            @PathVariable String trustId) {
+
+        DataMap dataMap = new DataMap.Builder().build();
+        dataMap.getLogMap().put("trust_id", trustId);
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesController.java:29]: we are supposed to use the builder. I suspect there is no trustId key pre-defined so in this case maybe just have the trust id in the main log message. Which could well be an info one.

File: src/main/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesController.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        LOGGER.debugContext(trustId,
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesController.java:30]: infoConect(

File: src/main/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesController.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                "Calling service to retrieve Corporate Trustee Data for Trust Id " + trustId,
+                dataMap.getLogMap());
+
+        try {
+            List<CorporateTrusteeData> corporateTrusteeData = corporateTrusteesService.getCorporateTrusteeData(
+                    trustId);
+            LOGGER.debugContext(trustId, "Returning Corporate Trustee Data for Trust Id " + trustId,
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesController.java:37]: info context

File: src/main/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesController.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                    dataMap.getLogMap());
+
+            return ResponseEntity.status(HttpStatus.OK).body(corporateTrusteeData);
+        } catch (TrustDataCountNotFoundException e) {
+
+            LOGGER.debugContext(trustId,
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesController.java:43]: infoContext

File: src/main/java/uk/gov/ch/model/update/trusts/CorporateTrusteeData.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    private String trusteeName;
+
+    @Column(name = "REGISTER_LOCATION")
+    @JsonProperty("registerLocation")
+    private String registerLocation;
+
+    @Column(name = "REGISTRATION_NUMBER")
+    @JsonProperty("registrationNumber")
+    private String registrationNumber;
+
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/ch/model/update/trusts/CorporateTrusteeData.java:26]: @Column(name="country")
@JsonProperty("country")
private String country;

File: src/main/java/uk/gov/ch/model/update/trusts/CorporateTrusteeData.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    @JsonProperty("legalForm")
+    private String legalForm;
+
+    @Column(name = "ON_REGISTER_IN_COUNTRY_FORMED")
+    @JsonProperty("onRegisterInCountryFormed")
+    private Boolean onRegisterInCountryFormed;
+
+
+    @Column(name = "CORPORATE_IND")
+    @JsonProperty("corporateInd")
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/ch/model/update/trusts/CorporateTrusteeData.java:41]: corporateIndicator

File: src/main/java/uk/gov/ch/repository/update/trusts/CorporateTrusteesRepository.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+        PagingAndSortingRepository<CorporateTrusteeData, Long> {
+    @Query(value = "SELECT ctd.trustee_id, ctd.trustee_name, "
+            + "ctd.on_register_in_country_formed, ctd.register_location, ctd.registration_number, "
+            + "ctd.law_governed, ctd.legal_form,"
+            + "t.corporate_ind, t.trustee_type_id,"
+            + "t.appointment_date, t.ceased_date"
+            + "FROM trustee t"
+            + "JOIN corporate_trustee_details ctd "
+            + "ON t.trustee_id = ctd.trustee_id "
+            + "WHERE t.trust_id = '?'",
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/ch/repository/update/trusts/CorporateTrusteesRepository.java:18]: "where t.trust_id = ?"

File: src/main/java/uk/gov/ch/service/update/trusts/impl/CorporateTrusteesServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        List<CorporateTrusteeData> corporateTrusteeDataList = corporateTrusteesRepository.getCorporateTrustees(
+                trustId);
+
+
+        if (corporateTrusteeDataList == null || corporateTrusteeDataList.isEmpty()) {
+            throw new TrustDataCountNotFoundException(
+                    "No corporate trustee data found with trustId " + trustId);
+        } else {
+            DataMap dataMap = new DataMap.Builder().build();
+            dataMap.getLogMap().put("trust_id", trustId);
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/ch/service/update/trusts/impl/CorporateTrusteesServiceImpl.java:37]: We should be using the new builder for log maps. But there probably is no key for trustIds so maybe just include in log message.

File: src/main/java/uk/gov/ch/model/update/trusts/CorporateTrusteeData.java (hunk 10 above)
Comment (AndreKramer7654) [reply src/main/java/uk/gov/ch/model/update/trusts/CorporateTrusteeData.java:26]: yes please - must have missed it. Maybe mistaken for address.

Comment (RobertBromfield) [reply src/main/java/uk/gov/ch/model/update/trusts/CorporateTrusteeData.java:26]: Done

File: src/test/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        setTrusteeId(123L);
+        setTrusteeName("My trust");
+        setRegisterLocation("London, UK");
+        setRegistrationNumber("R12345");
+        setLawGoverned("UK Law");
+        setLegalForm("LLP");
+        setOnRegisterInCountryFormed(true);
+        setCorporateInd("Y");
+        setTrusteeTypeId(1L);
+        setAppointmentDate("2023-01-01"); // Note: This way of constructing Date is deprecated; consider using LocalDate and adjust your entity accordingly.
```
Comment (bbell-companieshouse) [inline src/test/java/uk/gov/ch/controller/update/trusts/CorporateTrusteesControllerTest.java:41]: Remove this comment

File: src/main/java/uk/gov/ch/model/update/trusts/CorporateTrusteeData.java [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+import java.util.Objects;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+
+@Entity
+public class CorporateTrusteeData {
+
+    @Column(name = "TRUSTEE_ID")
+    @JsonProperty("trusteeId")
+    private Long trusteeId;
```
[the remark this answers is not in the document]
Comment (RobertBromfield) [reply src/main/java/uk/gov/ch/model/update/trusts/CorporateTrusteeData.java:13]: As I said on the other PR I don't see a reason to and it's likely to be less efficient.

They're a number in the database so should really be a number in the model

Storing everything as String is really bad practice and if nothing else can lead us to the issues we've been having with Dates.

Though to be fair I doubt that's possible here I still think it's probably a bad idea

**Review by bbell-companieshouse — APPROVED** [review]
